libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc

```diff
@@ -0,0 +1,366 @@
+
+template<typename Fp12T>
+Fp12_2over3over2_square_gadget<Fp12T>::Fp12_2over3over2_square_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp12_2over3over2_variable<Fp12T> &A,
+    const Fp12_2over3over2_variable<Fp12T> &result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _A(A)
+    , _result(result)
+    , _alpha(
+          pb,
+          _A._c0,
+          _A._c1,
+          _result._c1 * (FieldT("2").inverse()),
+          FMT(annotation_prefix, " a0*a1"))
+    , _beta(
+          pb,
+          _A._c0 + _A._c1,
+          _A._c0 +
+              mul_by_non_residue(pb, _A._c1, FMT(annotation_prefix, " a1*v")),
+          _result._c0 +
+              mul_by_non_residue(
+                  pb, _alpha._result, FMT(annotation_prefix, " alpha*v")) +
+              _alpha._result,
+          FMT(annotation_prefix, " (a0+a1)(a0+v.a1)"))
+{
+}
+
+template<typename Fp12T>
+const Fp12_2over3over2_variable<Fp12T>
+    &Fp12_2over3over2_square_gadget<Fp12T>::result() const
+{
+    return _result;
+}
+
+template<typename Fp12T>
+void Fp12_2over3over2_square_gadget<Fp12T>::generate_r1cs_constraints()
+{
+    _alpha.generate_r1cs_constraints();
+    _beta.generate_r1cs_constraints();
+}
+
+template<typename Fp12T>
+void Fp12_2over3over2_square_gadget<Fp12T>::generate_r1cs_witness()
+{
+    const Fp6T a0 = _A._c0.get_element();
+    const Fp6T a1 = _A._c1.get_element();
+    const Fp6T alpha = a0 * a1;
+    _result._c1.generate_r1cs_witness(alpha + alpha);
+    _alpha.generate_r1cs_witness();
+
+    const Fp6T beta = (a0 + a1) * (a0 + Fp12T::mul_by_non_residue(a1));
+    _result._c0.generate_r1cs_witness(
+        beta - Fp12T::mul_by_non_residue(alpha) - alpha);
+    _beta._A.evaluate();
+    _beta._B.evaluate();
+    _beta.generate_r1cs_witness();
+}
+
+template<typename Fp12T>
+Fp6_3over2_variable<typename Fp12T::my_Fp6> Fp12_2over3over2_square_gadget<
+    Fp12T>::
+    mul_by_non_residue(
+        libsnark::protoboard<FieldT> &pb,
+        const Fp6_3over2_variable<typename Fp12T::my_Fp6> &c,
+        const std::string &annotation_prefix)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        pb, c._c2 * Fp12T::non_residue, c._c0, c._c1, annotation_prefix);
+}
+
+// Fp12_2over3over2_mul_by_024_gadget methods
+
+template<typename Fp12T>
+Fp12_2over3over2_mul_by_024_gadget<Fp12T>::Fp12_2over3over2_mul_by_024_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp12_2over3over2_variable<Fp12T> &Z,
+    const libsnark::Fp2_variable<Fp2T> &X_0,
+    const libsnark::Fp2_variable<Fp2T> &X_2,
+    const libsnark::Fp2_variable<Fp2T> &X_4,
+    const Fp12_2over3over2_variable<Fp12T> &result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _Z(Z)
+    , _X_0(X_0)
+    , _X_2(X_2)
+    , _X_4(X_4)
+    // out_z0 = z0*x0 + non_residue * ( z1*x2 + z4*x4 )
+    // => z0 * x0 = out_z0 - non_residue * ( z1*x2 + z4*x4 )
+    , _z1_x2(
```

Comment:
A note about the form of the gadgets here ...

I think we remarked in the past that there were a lot of gadgets with a `shared_ptr` to another gadget that they seemingly "owned" (i.e. there was no reason for a pointer, especially not a smart pointer and especially not a ref-counted smart pointer).  While working on this code, I realised that in many cases smart pointers were used so that these "owned" gadgets or "sub-gadgets" could be constructed in code (without needing to keep a default-constructed uninitialized version around).

In these gadgets, the sub-gadgets are members of the larger parent gadget and are initialized when the parent gadget is constructed (rather than in code).  This creates these big member initialization statements in some places, but removes the need for smart pointers (in most cases) and the code that allocates members and sets up the smart pointers.